math/smatrix/inc/Math/CholeskyDecomp.h

```diff
@@ -312,14 +323,43 @@ template<class F> class CholeskyDecompGenDim
 private:
    /** @brief dimensionality
     * dimensionality of the problem */
-   unsigned fN;
+   unsigned fN = 0;
    /// lower triangular matrix L
    /** lower triangular matrix L, packed storage, with diagonal
     * elements pre-inverted */
-   F *fL;
+   std::unique_ptr<F[]> fL;
    /// flag indicating a successful decomposition
-   bool fOk;
+   bool fOk = false;
 public:
+   // default construction, move construction and assignment, and destruction
+   // do the right thing, so use the versions that the compiler generates for
+   // us...
+   CholeskyDecompGenDim() = default;
+   CholeskyDecompGenDim(CholeskyDecompGenDim&&) = default;
+   CholeskyDecompGenDim& operator=(CholeskyDecompGenDim&&) = default;
+   ~CholeskyDecompGenDim() = default;
+
+   /// copy constructor
+   CholeskyDecompGenDim(const CholeskyDecompGenDim& other)
+           : fN(other.fN), fL(std::make_unique<F[]>((fN * (fN + 1)) / 2)),
+             fOk(other.fOk)
+   {
+       std::copy(other.fL, other.fL + (fN * (fN + 1)) / 2, fL);
+   }
+   /// (copy) assignment
+   CholeskyDecompGenDim& operator=(const CholeskyDecompGenDim& other)
+   {
+       if (this != &other) {
+           // check if fL is large enough to hold other, if not, reallocate
+           if (fN < other.fN)
+               fL = std::make_unique<F[]>((other.fN * (other.fN + 1)) / 2);
+           fN = other.fN;
+           std::copy(other.fL, other.fL + (fN * (fN + 1)) / 2, fL);
+           fOk = other.fOk;
+       }
+       return *this;
+   }
+
```

Comment:
```suggestion
```
With the `std::vector` you wouldn't need any of that. Everything would be implicitly defaulted to the right behavior.

Also, why did you add a constructor with no arguments here? There was no such default constructor before. What would be the usecase of that feature?

Reply:
Please let me check if things like std::optional in gcc's C++ standard lib and LLVM's libc++ standard library deal well without the default constructor (the idea was that you can default-construct a CholeskyDecompGenDim without paying for anything as expensive as a memory allocation, and then move to the default-constructed one if all goes well). If I may, I'll get back to you once I've checked my use case.

Reply:
Okay, I checked both gcc's C++ standard library and LLVM's libc++, and it appears I don't need the default constructor after all. (Probably I did at some intermediate stage in my debugging and development session on Thursday night... ;) ) I'll prepare a patch to remove that one.